elastic/kibana
# [Reporting/Telemetry] Include counts for each deprecated job type

Closes https://github.com/elastic/kibana/issues/105868

## Summary

This PR adds an explicit `deprecated` count to each export type, which makes deprecating code in Reporting more flexible for telemetry:
 - you don't have to know which export types are deprecated to find usage of deprecated reporting features
 - you can compare overall deprecated count with overall report count.

Telemetry data can show how often deprecated export types are used. Prior to this PR, the export type itself would have to be deprecated, and we would find the usage of deprecated export types by comparing the usage of each export type.

When is merged, the `printable_pdf` and `PNG`, report jobs will be conditionally marked as deprecated, based on whether the job parameters used the 6.2 `savedObjectId` style, or the later `relativeUrls` style. Since the job type can not be used to see if the job is deprecated, this PR adds extra logic to telemetry collection in Reporting and gives an explicit flag on if the job is deprecated.

## Testing:
1. In Kibana 7.12, import test data, create a saved search in Discover, and copy the  POST URL for CSV export.
1. Create a script that sends the POST URL to Kibana to request a CSV report. Example: https://gist.github.com/tsullivan/0b86d5e1165f6460a876b4f63012fdd0
2. Create a few report jobs using the script
3. Export the saved objects in **Stack Management > Saved Objects**
3. Upgrade ES to 8.0, or import the test data and saved objects in ES 8.0 with running Kibana from source checked out to this PR.
2. Create a few more report jobs using the script
4. Run the script that uses the POST URL copied in 7.12.
  
  With the reporting index set up with deprecated job types, you can then query the internal API for usage, and see that it contains info about the deprecated job types.

7. Run a curl command on the usage endpoint and pipe the response data to `jq`:
 ```
  curl 'http://elastic:changeme@localhost:5601/api/telemetry/v2/clusters/_stats' --compressed \
   -H 'Content-Type: application/json' \
   -H 'kbn-version: 8.0.0' \
   --data-raw '{"unencrypted":true}' \
   | jq '.[0].stack_stats.kibana.plugins.reporting|{l7D:.last7Days|{csv,csv_searchsource},csv,csv_searchsource}'
 ```
  Example output:
  ```
{
  "l7D": {
    "csv": {
      "available": true,
      "total": 4,
      "deprecated": 4
    },
    "csv_searchsource": {
      "available": true,
      "total": 5,
      "deprecated": 0
    }
  },
  "csv": {
    "available": true,
    "total": 4,
    "deprecated": 4
  },
  "csv_searchsource": {
    "available": true,
    "total": 5,
    "deprecated": 0
  }
}
```

## TODO:
 - [x] Mark all count of a job type as deprecated, if the job type itself is deprecated

### Checklist

Delete any items that are not applicable to this PR.
